chain-signatures/node/src/protocol/signature.rs

```diff
@@ -1224,51 +1340,52 @@ pub struct SignatureSpawnerTask {
+    /// The spawner will wait for the contract to reach Running state before processing requests.
+    /// It dynamically obtains governance info (me, threshold, epoch, public_key) from the contract
+    /// state and updates when the contract transitions to a new epoch after resharing.
+    #[allow(clippy::too_many_arguments)]
     pub fn run(
-        me: Participant,
-        threshold: usize,
-        epoch: u64,
-        ctx: &MpcSignProtocol,
-        public_key: PublicKey,
+        my_account_id: AccountId,
+        sign_rx: mpsc::Receiver<Sign>,
+        contract: ContractStateWatcher,
+        config: watch::Receiver<Config>,
+        presignature_storage: PresignatureStorage,
+        mesh_state: watch::Receiver<MeshState>,
+        msg_channel: MessageChannel,
+        rpc_channel: RpcChannel,
+        backlog: Backlog,
     ) -> Self {
         let spawner = SignatureSpawner {
-            me,
             tasks: JoinMap::new(),
             inboxes: HashMap::new(),
-            my_account_id: ctx.my_account_id.clone(),
-            threshold,
-            public_key,
-            epoch,
-            presignatures: ctx.presignature_storage.clone(),
-            mesh_state: ctx.mesh_state.clone(),
-            msg: ctx.msg_channel.clone(),
-            rpc: ctx.rpc_channel.clone(),
-            backlog: ctx.backlog.clone(),
+            my_account_id,
+            presignatures: presignature_storage,
+            mesh_state,
+            contract: contract.clone(),
+            msg: msg_channel,
+            rpc: rpc_channel,
+            backlog,
         };
 
         Self {
-            handle: tokio::spawn(spawner.run(
-                ctx.sign_rx.clone(),
-                ctx.contract.clone(),
-                ctx.config.clone(),
-            )),
+            handle: tokio::spawn(spawner.run(sign_rx, contract, config)),
         }
     }
 
     pub fn abort(&self) {
-        // NOTE: since dropping the handle here, PresignatureSpawner will drop their JoinSet/JoinMap
-        // which will also abort all ongoing presignature generation tasks. This is important to note
-        // since we do not want to leak any presignature generation tasks when we are resharing, and
-        // potentially wasting compute.
+        // NOTE: This aborts the spawner task and all ongoing signature generation tasks.
+        // This is typically only called during shutdown or catastrophic error recovery.
+        // During normal operation (including resharing), the spawner should NOT be aborted
+        // since sign tasks need to persist across resharing.
         self.handle.abort();
     }
 }
 
-impl Drop for SignatureSpawnerTask {
-    fn drop(&mut self) {
-        self.abort();
-    }
-}
+// NOTE: We intentionally do NOT implement Drop for SignatureSpawnerTask.
+// The spawner should persist across resharing, so dropping it should not abort the task.
+// The spawner will naturally terminate when all channels close (e.g., during node shutdown).
```

Comment:
Wait, I don't quite follow, when would `SignatureSpawnerTask` drop anyway? It's stored in `MpcSignProtocol`, right? So the only way it would be dropped is if `MpcSignProtocol` drops. And that's exactly when you call abort anyway.

I don't understand the difference between what you have now and aborting the handle whenever `SignatureSpawnerTask` drops. What am I missing? :confused: 

Also, could you put this comment somewhere closer to the `SignatureSpawnerTask` definition? Or inside `fn abort` just above? Seems a but out of place here and I probably wouldn't read it when I look at the relevant other code.